#### Ashraya Joshi
#### CSCI 692 - Randomized Algorithms
#### Project 1 - Simulated Annealing

We are given a .png file that represents columns of a data. The objective is to find the optimal permutations of the image. The problem is closely related to the "Traveling Salesman Problem".

In [67]:
# Import the image using Image and ImageOps module from PIL package.
from PIL import Image, ImageOps
import numpy as np
from scipy import spatial
import cv2
import math
# Create an image object and convert to 8-bit grayscale image
img = Image.open("test.png").convert("L")
im_gray = cv2.imread("test.png", cv2.IMREAD_GRAYSCALE)
WIDTH, HEIGHT = img.size
CITY_COUNT = WIDTH
INITIAL_TEMPERATURE = 100000
STOPPING_TEMPERATURE = 1
TEMPERATURE_DECAY = 0.9995
img.size[0]

1522

In [68]:
cities = np.array(img)
cities = cities.T
# print(np.linalg.norm(cities[:, 0] - cities[:, 1]))
# print(np.sqrt(np.sum(np.square(cities[:, 0] - cities[:, 1]))))

So, we have 1522 columns in the matrix, i.e. 1522 columns in the image. If we treat the problem as a TSP, we have 1522 cities. First, we will change every column of the matrix to their specific vectors.

In [69]:
def length(n1, n2):
    return np.sqrt(np.sum(np.square(cities[n1] - cities[n2])))

In [70]:
def total_length(cities, n):
    l = length(cities[0], cities[n-1])
    for i in range(n-1):
        l += length(cities[i], cities[i + 1])
    return l

In [73]:
def two_opt_optimization(sol_arr, t, n):
    ai = np.random.randint(0, n-1)
    bi = (ai + 1) % n
    ci = np.random.randint(0, n-1)
    di = (ci + 1) % n
    
    if ai != ci and bi != ci:
        a = sol_arr[ai]
        b = sol_arr[bi]
        c = sol_arr[ci]
        d = sol_arr[di]
        
        ab = length(a, b)
        cd = length(c, d)
        
        ac = length(a, c)
        bd = length(b, d)
        
        diff = (ab + cd) - (ac + bd)
        
        p = 0
        
        if diff < 0:
            p = math.exp(diff / t)
        elif diff > 0.05:
            p = 1
            
        if np.random.random() < p:
            new_arr = list(range(0, n))
            new_arr[0] = sol_arr[ai]
            i = 1
            
            while bi != ci:
                new_arr[i] = sol_arr[ci]
                i = i + 1
                ci = (ci - 1) % n
                
            new_arr[i] = sol_arr[bi]
            i = i + 1
            
            while ai != di:
                new_arr[i] = sol_arr[di]
                i = i + 1
                di = (di + 1) % n
                
            return new_arr
    return sol_arr
        

In [74]:
def sa_algorithm(cities):
    n = CITY_COUNT
    
    sol_arr = cities
    
    t = 100
    min_l = total_length(sol_arr, n)
    
    i = 0
    best_arr = []
    
    while t > 0.1:
        i = i + 1
        sol_arr = two_opt_optimization(sol_arr, t, n)
        
        if i >= 200:
            i = 0
            current_l = total_length(sol_arr, n)
            
            t = t * 0.9995
            
            if current_l < min_l:
                print(current_l)
                min_l = current_l
                best_arr = sol_arr[:]
    return best_arr

final_arr = sa_algorithm(cities)
final_l = total_length(final_arr, n)

7719669.395254193
7485267.463031678
7289422.681408672
7099853.388917846
6929475.279296251
6774672.81804115
6655696.156918028
6479869.345264945
6373701.027568095
6264533.538285743
6130627.9888084335
6023657.6404277515
5914213.959866108
5852335.589380155
5742211.560590372
5665715.751382618
5578088.165192427
5528618.603698223
5467790.220957929
5414777.699890931
5360791.194330225
5283796.316578015
5229410.518423092
5169414.072824739
5119482.560172402
5082410.946020366
5034037.484019711
4959480.363041541
4909498.019324937
4864610.682473227
4832980.996983936
4784814.3059635125
4754391.585317557
4701694.538637949
4680615.864386678
4654710.086332129
4606657.394182729
4579596.489115586
4546487.936254414
4510732.404317462
4485701.440537374
4452774.60279433
4436977.019865222
4395566.737449031
4353116.390396047
4335388.702258411
4314382.292353227
4289692.001489836
4248372.840068828
4226322.504945459
4218089.12942837
4195244.9268243145
4175768.253810459
4156265.032864936
4134736.95411498
4118114.01

2341044.3353517703
2338700.915390212
2338518.3481638497
2337160.9898357242
2335924.5166953104
2335105.467146404
2331817.9377013496
2331107.404336622
2329417.4150464265
2327790.32289342
2326105.0669929385
2323875.892013824
2321671.603058412
2321339.2926250505
2319983.444837765
2317605.5184982726
2316674.7332173176
2316186.639337856
2315595.4468205534
2315193.995765962
2314686.863126078
2313945.30881165
2313466.756058272
2313057.282871555
2312978.9427253255
2312861.074875475
2311850.5046483213
2310108.425326047
2308084.506492874
2307761.167500157
2307595.3999925433
2305053.973615502
2303695.697901533
2303612.9649695437
2302429.3127390617
2301942.5838952404
2300220.379192479
2299721.0413928144
2299602.40536718
2298367.846397774
2296425.7796589094
2293986.549651689
2293764.2179573323
2293098.898556854
2292885.5725461096
2291548.2297684075
2291093.8310629153
2289131.339709476
2286469.045457507
2285660.653592672
2282677.0305426745
2281806.9029705734
2280564.076042636
2279128.0587729677
22780

KeyboardInterrupt: 

In [75]:
print(final_l)

NameError: name 'final_l' is not defined

In [31]:
def Initialize(count):
    solution = np.arange(count)
    np.random.shuffle(solution)
    return solution

In [32]:
def Evaluate(cities, solution):
    distance = 0
    for i in range(cities.shape[1]):
        index_a = solution[i]
        index_b = solution[i - 1]
        distance += np.sqrt(np.sum(np.square(cities[:, index_a] - cities[:, index_b])))
    return distance

In [33]:
def Modify(current):
    new = current.copy()
    index_a = np.random.randint(len(current))
    index_b = np.random.randint(len(current))
    while index_b == index_a:
        index_b = np.random.randint(len(current))
    new[index_a], new[index_b] = new[index_b], new[index_a]
    return new

In [41]:
current_solution = Initialize(CITY_COUNT)
print(current_solution)
current_score = Evaluate(cities, current_solution)
print(current_score)
best_score = worst_score = current_score
temperature = INITIAL_TEMPERATURE
while temperature > STOPPING_TEMPERATURE:
    new_solution = Modify(current_solution)
    new_score = Evaluate(cities, new_solution)
    best_score = min(best_score, new_score)
    worst_score = max(worst_score, new_score)
    if new_score < current_score:
        current_solution = new_solution
        current_score = new_score
    else:
        delta = abs(new_score - current_score)
        probability = np.exp(-delta / temperature)
        if probability > np.random.uniform():
            current_solution = new_solution
            current_score = new_score
    temperature *= TEMPERATURE_DECAY

[ 932 1239 1367 ... 1001 1090   89]
281087.01060841454


In [49]:
current_solution

array([ 962, 1307,  201, ...,  502,  327,  257])

In [50]:
final_matrix = []
for index in current_solution:
    final_matrix.append(cities[:, index])
final_matrix = np.vstack(final_matrix)
print(final_matrix.T.shape)
final_image = Image.fromarray(final_matrix.T, "L")
final_image.show()

(1395, 1522)
